syntax = "proto3";

package cosyvoice;
option go_package = "protos/";

service CosyVoice{
  rpc Inference(Request) returns (stream Response) {}
  rpc SaveShot(SaveShotRequest) returns (SaveShotResponse) {}
  rpc AppZeroOutput(AppZeroOutputRequest) returns (stream Response) {}
  rpc InferenceCrossLingual(CrossLingualRequest) returns (stream Response) {}
  rpc Translate(TranslateRequest) returns (TranslateResponse) {}
}

message Request{
  oneof RequestPayload {
    sftRequest sft_request = 1;
    zeroshotRequest zero_shot_request = 2;
    crosslingualRequest cross_lingual_request = 3;
    instructRequest instruct_request = 4;
  }
}

message sftRequest{
  string spk_id = 1;
  string tts_text = 2;
}

message zeroshotRequest{
  string tts_text = 1;
  string prompt_text = 2;
  bytes prompt_audio = 3;
}

message crosslingualRequest{
  string tts_text = 1;
  bytes prompt_audio = 2;
}

message instructRequest{
  string tts_text = 1;
  string spk_id = 2;
  string instruct_text = 3;
}

message Response{
  bytes tts_audio = 1;
}

message SaveShotRequest {
  bytes prompt_audio = 1;
  string file_name = 2;
}

message SaveShotResponse {
  bool success = 1;
}

message AppZeroOutputRequest {
  string file_name = 1;
  string tts_text = 2;
  string prompt_text = 3;
  string user_id = 4;
  bool stream = 5;
  float speed = 6;
  int32 seed = 7;
}

message CrossLingualRequest {
  string file_name = 1;
  string tts_text = 2;
  string user_id = 3;
  bool stream = 4;
  float speed = 5;
  int32 seed = 6;
}

message TranslateRequest {
  string text = 1;
}

message TranslateResponse {
  string translated_text = 1;
}